### PR TITLE
Viking tokenizer support

### PR DESCRIPTION
**Pull request by akx**
[LumiOpen/Viking-7B](https://huggingface.co/LumiOpen/Viking-7B) has a variant tokenizer

Just using `llama-bpe` had the model generate sensible Finnish, but this is attempting to do things a bit more correctly.

See 

* https://huggingface.co/LumiOpen/Viking-7B/discussions/2#664602c8b88f8519c3d50113 for discussion.
* converted model (with just `llama-bpe`): https://huggingface.co/akx/Viking-7B-gguf

### CODE REVIEW

**Review by ggerganov — COMMENTED** [review]
The proper way is to update `convert-hf-to-gguf-update.py` and validate that the tokenization tests pass

**Comment by akx** [comment]
@ggerganov Thanks for the pointers. I think I'm doing things more correctly in this iteration, but the `test-tokenizer-0` test is failing ([output gist here](https://gist.github.com/akx/bdde8c2184f6238e490e38b30620c650)), and I'm not quite sure where to go from there, even after trying to follow #6920...

`test-tokenizer-1-bpe` doesn't fail, but prints

> llm_load_vocab: mismatch in special tokens definition ( 11/131072 vs 24/131072 ).

among other output.

cc @jonabur (of the Viking team)

**Comment by ggerganov** [comment]
If `test-tokenizer-0` fails, this likely means that the pre-tokenizer config is not exactly the same as LLaMA3. I just checked the `tokenizer.json` file of that model and here is the relevant section:

```json
  "pre_tokenizer": {
    "type": "Sequence",
    "pretokenizers": [
      {
        "type": "Split",
        "pattern": {
          "Regex": " ?[^(\\s|[.,!?…。，、।۔،])]+"
        },
        "behavior": "Isolated",
        "invert": false
      },
      {
        "type": "Digits",
        "individual_digits": true
      },
      {
        "type": "ByteLevel",
        "add_prefix_space": false,
        "trim_offsets": true,
        "use_regex": false
      }
    ]
  },
```

So you have to implement use the respective regexes in `llama.cpp` instead of re-using LLaMA3:

https://github.com/ggerganov/llama.cpp/blob/27b040691cbe45314147c2745e891a38e9c048d4/llama.cpp#L12287-L12378

**Comment by jonabur** [comment]
We've also run into this same issue providing a gguf'd version of Poro, using the same tokenizer regex.  @akx have you made any progress converting the regex format?

**Comment by akx** [comment]
I'm not sure what to make of the failing tokenizer tests even after adding in the regexp (and I suppose "digits" means another regexp should be splitting all digits into separate tokens?)

The actual detokenized output matches the expected output, but the token sequence doesn't. The first commit in this PR now improves the output of `test-tokenizer-0` for easier diffing...

```diff
-expected tokens:
+got tokens:
    746 '
  '
   2392 '
 
  '
  55899 '
@@ -169,37 +169,43 @@
   3395 '""'
  30917 '......'
  17846 '!!!!'
   2420 '!!'
  13728 '????'
   3963 '??'
-  9873 ' I've'
+   383 ' I'
+  7029 ''ve'
   1912 ' been'
- 37493 ' 't'
+   630 ' ''
+   107 't'
    733 'old'
- 17600 ' he's'
+   627 ' he'
+   689 ''s'
   1923 ' there'
     35 ','
    630 ' ''
   1417 'RE'
    791 ' you'
   6189 ' sure'
     54 '?'
- 23586 ' 'M'
+   630 ' ''
+    68 'M'
    835 ' not'
   6189 ' sure'
- 18068 ' I'll'
+   383 ' I'
+  6704 ''ll'
   2463 ' make'
    590 ' it'
     35 ','
- 35018 ' 'D'
+   630 ' ''
+    59 'D'
    791 ' you'
   1647 ' like'
   2032 ' some'
  22940 ' tea'
     54 '?'
   2221 ' We'
     30 '''
   6815 'Ve'
    279 ' a'
  79905 ''l'
     67 'L'
```
and
```diff
-expected tokens:
+got tokens:
    348 '   '
  40540 ' Hello'
-   472 '
+   209 '
    '
+   348 '   '
  40540 ' Hello'
```

**Comment by jonabur** [comment]
I'm not familiar with the tokenizer regex unfortunately, we inherited it from Bloom (maybe it should be called bloom tokenizer, instead?) but it looks like it's using some possibly non-standard regex features, and may not translate directly to the regex format supported in llama.cpp.

In particular capturing groups inside character classes, or embedding character classes inside character classes both seem possibly non-standard to me, though it's been a long time since I've done regexes anywhere near this complicated.

File: llama.cpp [hunk 1]
```diff
@@ -4549,9 +4549,10 @@ static void llm_load_vocab(
                     tokenizer_pre == "default") {
                 vocab.type_pre = LLAMA_VOCAB_PRE_TYPE_DEFAULT;
             } else if (
-                    tokenizer_pre == "llama3"   ||
-                    tokenizer_pre == "llama-v3" ||
-                    tokenizer_pre == "llama-bpe") {
+                    tokenizer_pre == "llama3"    ||
+                    tokenizer_pre == "llama-v3"  ||
+                    tokenizer_pre == "llama-bpe" ||
+                    tokenizer_pre == "viking-7b") {
```
Comment (jonabur) [inline llama.cpp:4555]: This needs two changes:
  - tokenizer_pre needs to be updated to match the "viking" in convert-hf-to-gguf.py
  - this needs its own if statement block which sets vocab.type_pre = LLAMA_VOCAB_PRE_TYPE_VIKING

File: llama.cpp [hunk 2]
```diff
@@ -12580,6 +12581,11 @@ struct llm_tokenizer_bpe {
                             "(?:'[sS]|'[tT]|'[rR][eE]|'[vV][eE]|'[mM]|'[lL][lL]|'[dD])|[^\\r\\n\\p{L}\\p{N}]?\\p{L}+|\\p{N}| ?[^\\s\\p{L}\\p{N}]+[\\r\\n]*|\\s*[\\r\\n]+|\\s+(?!\\S)|\\s+",
                         });
                         break;
+                    case LLAMA_VOCAB_PRE_TYPE_VIKING:
+                        word_collection = unicode_regex_split(text, {
+                            " ?[^(\\s|[.,!?…。，、।۔،])]+",
```
Comment (jonabur) [inline llama.cpp:12586]: I think this regex works for the first test that fails, but I'm still left with one failing test I don't understand.

                            " ?[^\\s.,!?…。，、।۔،]+",

**Comment by jonabur** [comment]
I suggested two changes which enable the code to work and the first test to pass, but now a different test is passing and I don't understand why.  It looks like a unicode character is being split?  Any idea what's going on here?
```
src: 'ied 4 ½ months'
res: 'ied 4 ½ months'
tok: 1502 231 43 882 145 9290
failed test:    'ied 4 ½ months'
detokenized to: 'ied 4 ½ months'
(which matches the expected output)
expected tokens:
  1502 'ied'
   231 ' '
    43 '4'
   231 ' '
  1177 '½'
  9290 ' months'

got tokens:
  1502 'ied'
   231 ' '
    43 '4'
   882 ' �'
   145 '�'
  9290 ' months'
  ```

**Comment by ezosa** [comment]
Here’s a couple of updates related to this issue:

1. Poro-34b-chat pre-tokenizer support is now merged (#7713 ). Since Poro and Viking share the same pre-tokenizer regex, you can use _poro-chat_ as the pre-tokenizer type for Viking. So the changes to convert-hf-to-gguf.py will be:

```
        if chkhsh == "7fc505bd3104ca1083b150b17d088b59534ede9bde81f0dd2090967d7fe52cee":
            # ref: https://huggingface.co/LumiOpen/Viking-7B
            res = "poro-chat"
```

2. We looked further into the failed test for Viking and found that this block in tokenizer.json is related to the issue, under the _pre_tokenizer_ attribute (this does not exist in Poro's tokenizer so we didn't get this problem there):

```
      {
        "type": "Digits",
        "individual_digits": true
      },
```

If you remove this part and regenerate the vocab files and run the tests again, the tests will pass. 
The tokenization of the test strings with and without the digits block is the same. Therefore we think that removing that block will not cause any issues in the downstream performance of the quantised model.

This is the failed test:
```
src: 'ied 4 ½ months'
res: 'ied 4 ½ months'
tok: 1502 231 43 882 145 9290 
main : failed test:    'ied 4 ½ months'
main : detokenized to: 'ied 4 ½ months' instead of 'ied 4 ½ months'
main : expected tokens:   1502 'ied',    231 ' ',     43 '4',    231 ' ',   1177 '½',   9290 ' months', 
main : got tokens:        1502 'ied',    231 ' ',     43 '4',    882 ' �',    145 '�',   9290 ' months', 
```

After we remove the digits block in tokenizer.json and regenerate the vocab files and run the tests, we get:

```
src: 'ied 4 ½ months'
res: 'ied 4 ½ months'
tok: 1502 231 43 882 145 9290 
```

**Comment by ggerganov** [comment]
Try to add the following regex to the viking pre-tokenizer in `llama.cpp`:

```
"\\p{N}",
```

**Comment by ezosa** [comment]
Thanks, that worked! So now we need to separate the pre-tokenizers for Poro and Viking. In llama.cpp, this needs to be added:

```
            case LLAMA_VOCAB_PRE_TYPE_VIKING:
                regex_exprs = {
                    "\\p{N}",
                    " ?[^(\\s|.,!?…。，、।۔،)]+",
                };
                break;
```

**Comment by akx** [comment]
Closing in favor of #8135 :)

**Pull request closed without merging**